### PR TITLE
Relax singleton-context constraint for pending non-singleton-context reverts

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

When a migration runs with [-singleton-context](https://vitess.io/docs/13.0/user-guides/schema-changes/ddl-strategy-flags/) ddl strategy flag, and with context `A`, then another migration with `-singleton-context` can only be submitted if it has same context `A`, and cannot be submitted if it has context `B`. Note, the submission itself returns with an error.

We want to have better user experience when it comes to [reverts](https://vitess.io/docs/13.0/user-guides/schema-changes/revertible-migrations/). Reverts may be long running and are allowed to run concurrently to other migrations.

A REVERT can run without a `-singleton-context` flag. In fact, this is the canonical way of running it. So we do not fail submitting a REVERT even if there are pending migrations.

However, what happens when a REVERT is already running, and a new migration comes in?

Starting this PR, the following holds true:

- If a REVERT is running
- And it does **not** have `-singleton-context`
- Then a new migration can be submitted, even if it does have `-singleton-context` and with some arbitrary `migration_context`.

To clarify:
- If a REVERT is running
- And has a `-singleton-context` flag with some `-migration_context=A`
- Then a new migration that has a `-singleton-context` flag may only be submitted it it also has `-migration_context=A`, and it will be rejected if it has `-migration_context=B`

The reason we add this relaxation is that we see an operational pattern in usage, and this behavior makes sense operationally.

Documentation to follow.

## Related Issue(s)

- #6926 
- https://github.com/vitessio/vitess/pull/7946

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Call for review

File: go/vt/vttablet/onlineddl/executor.go [hunk 1]
```diff
@@ -3586,6 +3586,29 @@ func (e *Executor) CompleteMigration(ctx context.Context, uuid string) (result *
...
+		return true, nil
+	}
+
+	// The pending migration is a revert
+	if !pendingOnlineDDL.StrategySetting().IsSingletonContext() {
+		// Aha! So, our "conflict" is with a REVERT migration, which does _not_ have a -singleton-context
+		// flag. Because we want to allow REVERT migrations to run as concurrently as possible, we allow this scenario.
+		return false, nil
+	}
+	return true, nil
```
Comment (mattlord) [inline go/vt/vttablet/onlineddl/executor.go:3609]: What is the point of the error return? I don't see where that's used for anything as this is a binary/boolean call. Feels like it only complicates code on the caller side.

**Review by mattlord — COMMENTED** [review]
A couple of nits, but otherwise LGTM. We should at least change the single dashes to double for the new flag everywhere. Let me know what you think.

File: go/test/endtoend/onlineddl/singleton/onlineddl_singleton_test.go [hunk 2]
```diff
@@ -261,34 +264,59 @@ func TestSchemaChange(t *testing.T) {
...
 		uuids = append(uuids, uuid)
 		onlineddl.CheckMigrationStatus(t, &vtParams, shards, uuid, schema.OnlineDDLStatusComplete)
 		checkTable(t, tableName, true)
 	})
 
-	// Last two tests (we run an incomplete migration)
-	t.Run("submit successful migration, no wait, vtgate", func(t *testing.T) {
-		_ = testOnlineDDLStatement(t, alterTableTrivialStatement, "gh-ost -singleton", "vtgate", "hint_col", "", true)
+	t.Run("fail concurrent singleton, vtgate", func(t *testing.T) {
+		uuid := testOnlineDDLStatement(t, alterTableTrivialStatement, "vitess --postpone-completion -singleton", "vtgate", "", "hint_col", "", true)
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/test/endtoend/onlineddl/singleton/onlineddl_singleton_test.go:281]: those are actually not command line flags but `ddl_strategy` flags, so they're not subject to that rule; although I agree it's nce to be consistent everywhere.

File: go/vt/vttablet/onlineddl/executor.go [hunk 3]
```diff
@@ -3586,6 +3586,29 @@ func (e *Executor) CompleteMigration(ctx context.Context, uuid string) (result *
...
+func (e *Executor) submittedMigrationConflictsWithPendingMigrationInSingletonContext(
+	ctx context.Context, submittedMigration, pendingOnlineDDL *schema.OnlineDDL,
+) (bool, error) {
+	if pendingOnlineDDL.MigrationContext == submittedMigration.MigrationContext {
+		// same migration context. this is obviously allowed
+		return false, nil
+	}
+	// Let's see if the pending migration is a revert:
+	revertedUUID, _ := pendingOnlineDDL.GetRevertUUID()
+	if revertedUUID == "" {
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go]: changed to `if _, err := pendingOnlineDDL.GetRevertUUID(); err != nil ` ; I prefer not to rely on a specific error code.

File: go/vt/vttablet/onlineddl/executor.go (hunk 1 above)
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go]: Good catch! Must be a leftover from a previous flow. Removed `error` from return value

File: go/test/endtoend/onlineddl/singleton/onlineddl_singleton_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply go/test/endtoend/onlineddl/singleton/onlineddl_singleton_test.go]: Now using double dash in `ddl_strategy`

File: doc/releasenotes/14_0_0_summary.md [hunk 4]
```diff
@@ -64,6 +64,10 @@ Example:
 vtctlclient ApplySchema -skip_preflight -ddl_strategy='vitess' -sql "alter table my_table add column my_val int not null default 0" commerce
 ```
 
+### -singleton-context and REVERT migrations
```
[the remark this answers is not in the document]
Comment (shlomi-noach) [reply doc/releasenotes/14_0_0_summary.md]: fixed

**Review by mattlord — APPROVED** [review]
Thanks!

**Comment by shlomi-noach** [comment]
no no, thank _you_

**Pull request merged**